### PR TITLE
Ztunnel logging improvements

### PR DESCRIPTION
**Pull request by MorrisLaw**
**Please provide a description of this PR:**
Fixes Issue #44541.

- We were creating new Kube Clients every time `isZtunnelPod` was called. This PR changes that by initializing a Kube Client one time and using that for `isZtunnelPod`.
- This PR also adds a check during log parsing to ensure that we account for namespaces being added to the logger. Instead of invalidating valid logger level combinations, we now validate them properly.  **Below is a replication of the bug and an example of how it works now**:

**Note for Reviewer:** Formatting is off for the terminal output examples since it doesn't copy/paste easily to GitHub. Apologies in advance!

*Before fix*:
```
jmorris@jeremymorris:~/Dev/istio/istioctl/cmd/istioctl$ ./istioctl -n istio-system pc log ztunnel-8r6kp --level=ztunnel::pool:debug                                                                                                                                   
Error: unrecognized logging level: :pool:debug  
```

*After fix*:
```
jmorris@jeremymorris:~/Dev/istio/istioctl/cmd/istioctl$ ./istioctl -n istio-system pc log ztunnel-8r6kp --level=pool:debug                                              
ztunnel-8r6kp.istio-system: 
current log level is pool:debug=trace,info    
                                                                                                                          jmorris@jeremymorris:~/Dev/istio/istioctl/cmd/istioctl$ ./istioctl -n istio-system pc log ztunnel-8r6kp --level=ztunnel::pool:debug                                    
ztunnel-8r6kp.istio-system: 
current log level is ztunnel::pool:debug=trace,pool:debug=trace,info    
                                                                                                jmorris@jeremymorris:~/Dev/istio/istioctl/cmd/istioctl$ ./istioctl -n istio-system pc log ztunnel-8r6kp --level=test::ztunnel::pool:debug                               
ztunnel-8r6kp.istio-system:                                                        
current log level is test::ztunnel::pool:debug=trace,ztunnel::pool:debug=trace,pool:debug=trace,info         
                                                                                                                                               jmorris@jeremymorris:~/Dev/istio/istioctl/cmd/istioctl$ ./istioctl -n istio-system pc log ztunnel-8r6kp -r                                                              2023-05-20T11:13:02.543278Z     warn    unable to get logLevel from ConfigMap istio-sidecar-injector, using default value "warning" for envoy proxies and "info" for ztunnel                                                                                ztunnel-8r6kp.istio-system:                                                         current log level is info  
                                                                                                                                             jmorris@jeremymorris:~/Dev/istio/istioctl/cmd/istioctl$ ./istioctl -n istio-system pc log ztunnel-8r6kp --level=test::ztunnel::pool:debug                               
ztunnel-8r6kp.istio-system:                                                         
current log level is test::ztunnel::pool:debug=trace,info  
```

### CODE REVIEW

File: istioctl/cmd/proxyconfig.go [hunk 1]
```diff
@@ -116,13 +116,13 @@ var stringToLevel = map[string]Level{
 }
 
 var (
-	loggerLevelString = ""
-	reset             = false
+	loggerLevelString         = ""
+	reset                     = false
+	isZtunnelPodKubeClient, _ = kubeClient(kubeconfig, configContext)
```
Comment (hanxiaop) [inline istioctl/cmd/proxyconfig.go:121]: I think the error of `kubeClient` is checked across other commands. Could you also check the error instead of ignoring it?

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:121]: Hey @hanxiaop , I can do that if we want. It may be useful to note that the previous code did not check the error, I'm not sure if that was intended or not.

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:121]: Also, yes `kubeClient` is checked within this same file in every other instance except this one. It does seem more consistent to update this logic to also include an error check. I wonder if it still makes sense to bother abstracting it then :thinking:

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:121]: It looks like you authored the initial code with the ignored error check actually @hanxiaop 😅: https://github.com/istio/istio/commit/4721395704e91a388b8f33f296c64a5c77a906f6

Will update shortly!

Comment (keithmattix) [reply istioctl/cmd/proxyconfig.go:121]: I think an abstraction still makes sense; perhaps just wrap kubeClient initialization in something similar to a sync.Once?

File: istioctl/cmd/proxyconfig.go [hunk 2]
```diff
@@ -116,13 +118,21 @@ var stringToLevel = map[string]Level{
...
 )
 
 var isZtunnelPod = func(podName, podNamespace string) bool {
-	client, _ := kubeClient(kubeconfig, configContext)
-	return ambientutil.IsZtunnelPod(client, podName, podNamespace)
+	var err error
+	once.Do(func() {
+		isZtunnelPodKubeClient, err = kubeClient(kubeconfig, configContext)
+		if err != nil {
+			panic(err)
```
Comment (MorrisLaw) [inline istioctl/cmd/proxyconfig.go:132]: Let me know if a panic makes sense here or do we want to return? My initial gut is to avoid an extra value being returned `isZtunnelPod` because then we are doing error checks in every call to this and that seems less than ideal.

cc/ @keithmattix @hanxiaop

Comment (keithmattix) [reply istioctl/cmd/proxyconfig.go:132]: If you're going to panic, try and wrap the `kubeClient` call in some sort of retry (preferably pkg/backoff)

**Review by keithmattix — CHANGES_REQUESTED** [review]

**Review by keithmattix — APPROVED** [review]

**Comment by MorrisLaw** [comment]
@keithmattix 

Regarding:
[If you're going to panic, try and wrap the kubeClient call in some sort of retry (preferably pkg/backoff)](https://github.com/istio/istio/pull/45022#discussion_r1202610892)

Since ```// RetryWithContext tries the operation until it does not return error,
	// or when the context expires, whichever happens first.``` There seems to no longer be a need to `panic(err)` here.
	
When exponential backoff ends due to either expired context or returned error, the error value won't matter after the fact so if we care about the error surfacing, we should log it first. 

**TLDR;** I added the exponential backoff with logging instead of panic(err).

**Comment by keithmattix** [comment]
@MorrisLaw in that case, you want to add an aggressive timeout (~15sec) so that istioctl doesn't just hang forever.

File: istioctl/cmd/proxyconfig.go [hunk 3]
```diff
@@ -116,13 +120,26 @@ var stringToLevel = map[string]Level{
...
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+	var err error
+
+	b := backoff.NewExponentialBackOff(backoff.DefaultOption())
+	once.Do(func() {
+		err = b.RetryWithContext(ctx, func() error {
+			isZtunnelPodKubeClient, err = kubeClient(kubeconfig, configContext)
+			log.Warnf("failed to create k8s client")
+			return err
```
Comment (keithmattix) [inline istioctl/cmd/proxyconfig.go:139]: Just noticed this error got swallowed. You probably are better off just adding an error to the method signature; from a quick glance at callers, it looks like most can handle returning an error

**Review by keithmattix — CHANGES_REQUESTED** [review]

File: istioctl/cmd/proxyconfig.go [hunk 4]
```diff
@@ -847,7 +868,11 @@ func logCmd() *cobra.Command {
 				if err != nil {
 					return err
 				}
-				if isZtunnelPod(pod, podNamespace) {
+				ztunnelPod, err := isZtunnelPod(pod, podNamespace)
+				if err != nil {
+					return fmt.Errorf("failed to create k8s client: %v", err)
```
Comment (keithmattix) [inline istioctl/cmd/proxyconfig.go:873]: nit: the error message here is too specific; if there were other possible errors in isZtunnelPod, we wouldn't want to make the top-level message specific to the kubeClient.

**Review by keithmattix — APPROVED** [review]
One nit around error messages but LGTM

**Review by jaellio — APPROVED** [review]
LGTM

**Comment by zirain** [comment]
cc @hanxiaop

File: istioctl/cmd/proxyconfig.go [hunk 5]
```diff
@@ -457,8 +467,11 @@ func allConfigCmd() *cobra.Command {
 					if err != nil {
 						return err
 					}
-
-					if isZtunnelPod(podName, podNamespace) {
+					ztunnelPod, err := isZtunnelPod(podName, podNamespace)
```
Comment (hanxiaop) [inline istioctl/cmd/proxyconfig.go:470]: In `isZtunnelPod`, we initiate a new client, and in `extractConfigDump` and `extractZtunnelConfigDump`, we also initiate a new client to obtain the config dump. I think we can combine these two together.

Comment (hanxiaop) [reply istioctl/cmd/proxyconfig.go:470]: Maybe rename `isZtunnelPodKubeClient ` to `proxyConfigKubeClient` and use it everywhere in the proxyconfig command.

File: istioctl/cmd/proxyconfig.go (hunk 1 above)
Comment (hanxiaop) [reply istioctl/cmd/proxyconfig.go]: @MorrisLaw Thanks for catching the error. I think I made a mistake in the original code. I was considering whether having or not having the initial client error is acceptable, given that `isZtunnelPod` doesn't depend on the client because if not have, it will fallback to the name guess. However, even if it doesn't rely on the client, the subsequent part of extracing configdump, will still use the client.

File: istioctl/cmd/proxyconfig.go (hunk 5 above)
Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:470]: At first glance this seems a bit tricky because I'm not sure it's guaranteed if `proxyConfigKubeClient` is set for each of these instances where we create a kube client.

Right now it's only set when `isZtunnelPod` is called. And not every command calls `isZtunnelPod` first :thinking:

Going to have to find a way to populate that value first thing.. Maybe a few functions will still need to initialize but maybe they would just set the `proxyConfigKubeClient` value then.

I'll try some things and push up the change shortly.

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:470]: here's just one example of what I'm referring to:

```go
func workloadConfigCmd() {
.
.
.
    RunE: func(c *cobra.Command, args []string) error {
			var configWriter *ztunnelDump.ConfigWriter
			var err error
			if len(args) == 1 {
				if podName, podNamespace, err = getComponentPodName(args[0]); err != nil {
					return err
				}
				ztunnelPod, kubeClientErr := isZtunnelPod(podName, podNamespace)
				if kubeClientErr != nil {
					return fmt.Errorf("failed to determine if pod is a zTunnel pod: %v", kubeClientErr)
				}
				if !ztunnelPod {
					return fmt.Errorf("workloads command is only supported by ztunnel proxies: %v", podName)
				}
				configWriter, err = setupZtunnelConfigDumpWriter(podName, podNamespace, c.OutOrStdout())
			} else {
				configWriter, err = setupFileZtunnelConfigdumpWriter(configDumpFile, c.OutOrStdout())
			}
```

`getComponentPodName` calls `getPodNameWithNamespace` which currently initializes it's own kubeClient. Replacing that with `proxyConfigKubeClient` wouldn't work if `isZtunnelPod` wasn't called previously. Same issue if `len(args) != 1` and it goes down the path of calling `setupFileZtunnelConfigdumpWriter`.

Comment (keithmattix) [reply istioctl/cmd/proxyconfig.go:470]: If the goal is now to de-duplicate client inits in the entire package, you probably want a standalone method (getKubeClient) and a package-level sync.Once. Each of the functions you listed should call that instead of initializing its own kubeCLient

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:470]: @keithmattix 

Yeah I was thinking something along those lines:

```go
var (
	loggerLevelString     = ""
	reset                 = false
	once                  sync.Once
	proxyConfigKubeClient kube.CLIClient
)

var getProxyConfigKubeClient = func(kubeconfig, configContext string) error {
	var err error
	once.Do(func() {
		proxyConfigKubeClient, err = kubeClient(kubeconfig, configContext)
	})
	return err
}

var isZtunnelPod = func(podName, podNamespace string) (bool, error) {
	err := getProxyConfigKubeClient(kubeconfig, configContext)
	if err != nil {
		return false, err
	}
	return ambientutil.IsZtunnelPod(proxyConfigKubeClient, podName, podNamespace), nil
}
```

And callers would do similar to what isZtunnelPod is doing:
```go
err := getProxyConfigKubeClient(kubeconfig, configContext)
if err != nil {
	return "", fmt.Errorf("failed to create Kubernetes client: %v", err)
}
```
Then use `proxyConfigKubeClient` as needed.

I'm currently testing it now and it's failing unit tests. So I'm figuring that out atm.

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:470]: Ok, `package-level sync.Once` is implying that we need to do this across the entire package. I was interpreting their comment as just needing to do it in this file for this PR.

Does it make sense to couple the two goals `package level de-duplication` _and_ `log level validity improvement`? Seems like the scope is growing past the initial goal of the PR.

Comment (keithmattix) [reply istioctl/cmd/proxyconfig.go:470]: Good point; I'm in favor of a follow-up

Comment (hanxiaop) [reply istioctl/cmd/proxyconfig.go:470]: If the scope is expanding to the package level, I would prefer to do a follow-up as well.

**Review by hanxiaop — APPROVED** [review]
Thanks! LGTM

File: istioctl/cmd/proxyconfig.go [hunk 6]
```diff
@@ -116,13 +118,21 @@ var stringToLevel = map[string]Level{
...
+	reset                  = false
+	once                   sync.Once
+	isZtunnelPodKubeClient kube.CLIClient
 )
 
-var isZtunnelPod = func(podName, podNamespace string) bool {
-	client, _ := kubeClient(kubeconfig, configContext)
-	return ambientutil.IsZtunnelPod(client, podName, podNamespace)
+var isZtunnelPod = func(podName, podNamespace string) (bool, error) {
+	var err error
```
Comment (wulianglongrd) [inline istioctl/cmd/proxyconfig.go:128]: From the second call to `isZtunnelPod`, the `err` variable will always be nil.

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:128]: Yes

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:128]: Previous version never checked for an error on kube client initialization. This addresses that. And yeah, it's true it only happens once but that's because of the nature of once.Do(). Error will always be nil after the first call but we still need to ensure we error check for that first run where initialization of the kube client happens.

Does that make sense? Is there a more elegant way of handling that @wulianglongrd ?

Comment (wulianglongrd) [reply istioctl/cmd/proxyconfig.go:128]: I think that if you want to always hold the error of the first call, you can use a global variable, which is defined the same as `isZtunnelPodKubeClient`.

Comment (MorrisLaw) [reply istioctl/cmd/proxyconfig.go:128]: Good catch! I think this makes sense and has the added benefit of bringing `isZtunnelPod` back to being a one value return. Here's a PR: https://github.com/istio/istio/pull/45160

@wulianglongrd
